src/software/util/design_patterns/generic_factory.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+// TODO should be removed later
+#include <functional>
+//#include "software/backend/backend.h" // Should make generic.h
+
+
+// A quality of life typedef to make things shorter and more readable
+template <class GenericType>
+        using GenericRegistry = std::unordered_map<std::string, std::function<std::unique_ptr<GenericType>()>>;
+
+/**
+ * The GenericFactory is an Abstract class that provides an interface for Backend and Play
+ * Factories to follow. This makes it easy to maintain a list of factories and get the
+ * corresponding backends and plays through the generic interface.
+ */
+template <class GenericType>
```

Comment:
In general, I might suggest `TypeToBuild`, `TypeToCreate`, or `ReturnType`, just because `GenericType` is a little too.... _generic_ (please, hold your applause :sweat_smile:). Couldn't really think of a name I really like here though. @akhilveeraghanta @EvanMorcom @jonathanlew @MathewMacDougall for your suggestions? I'm really looking for something to replace "generic" everywhere in this class (comments and code).

Reply:
maybe `FactoryType`? Otherwise I like `TypeToCreate` the best.

Reply:
Moving forward with TypeToCreate.

Reply:
`TypeToCreate` is OK with me